mmu: Introduce arch_virt_region_align() for virt_region_alloc() aligment

superna9999 opened the pull request:
On ARM64 platforms, when mapping multiple memory zones with size
not multiple of a L2 block size (2MiB), all the following mappings
will probably use L3 tables.

And a huge mapping will consume all possible L3 tables.

In order to reduce usage of L3 tables, this introduces a new
arch_virt_region_align() optional architecture specific
call to eventually return a more optimal virtual address
alignment than the default MMU_PAGE_SIZE.

This alignment is used in virt_region_alloc() by:
- requesting more pages in virt_region_bitmap to make sure we request
  up to the possible aligned virtual address
- freeing the supplementary pages used for alignment

Add the arm64 MMU arch_virt_region_align() implementation used
to return a possible virtual addres alignment in order to
optimize the MMU table layout and possibly avoid using L3 tables
and use some L1 & L3 blocks instead for most of the mapping.

Suggested-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

npitre reviewed (commented):
I don't think a static Kconfig option is a good choice here. You risk
creating the opposite problem i.e. running out of virtual space after
a bunch of small mappings.

This should instead be a per-architecture macro or function that
provides at run time the best alignment to use given the physical
start address and the size, with a default implementation for
architectures that don't care.

superna9999 commented:
I don't see easy solutions for that, until a solution is found, mapping 256MiBs will need 130 L3 tables in the wrong init order.  (520KiB for 130 tables for an unaligned 256MiB mapping)

Nothing can predict mapping ordering at build-time, and it's a real issue when GICv3-ITS and PCIe ECAM are enabled on an ARM64 system.

As a short time solution, either you:
- greatly extend the MAX_XLAT_TABLES: but this will use true physical memory, 4KiB per table, which could be saved
- use this MMU_VIRT_ALIGN_SIZE trick: but only allow 2048 mappings in 32bit VA, but up to 134217728 in 48bit VA.
- statically change init order for drivers mapping multiple memory zones with size
not multiple of a L2 block size (2MiB) or multiple of L2 block size

npitre reviewed (commented):
This is still suboptimal and likely to waste vm space in some cases
if you don't take into account the physical address.

What would be needed is something like this:

```
diff --git a/kernel/mmu.c b/kernel/mmu.c
index 72f8bf9c1e..a35b463eeb 100644
--- a/kernel/mmu.c
+++ b/kernel/mmu.c
@@ -226,7 +226,7 @@ static void virt_region_init(void)
 	virt_region_inited = true;
 }
 
-static void *virt_region_alloc(size_t size)
+static void *virt_region_alloc(size_t size, size_t align)
 {
 	uintptr_t dest_addr;
 	size_t offset;
@@ -237,7 +237,7 @@ static void *virt_region_alloc(size_t size)
 		virt_region_init();
 	}
 
-	num_bits = size / CONFIG_MMU_PAGE_SIZE;
+	num_bits = (size + align - CONFIG_MMU_PAGE_SIZE) / CONFIG_MMU_PAGE_SIZE;
 	ret = sys_bitarray_alloc(&virt_region_bitmap, num_bits, &offset);
 	if (ret != 0) {
 		LOG_ERR("insufficient virtual address space (requested %zu)",
@@ -250,6 +250,11 @@ static void *virt_region_alloc(size_t size)
 	 * to get the starting address of the allocated region.
 	 */
 	dest_addr = virt_from_bitmap_offset(offset, size);
+	aligned_dest_addr = ROUND_UP(dest_addr, align);
+	if (aligned_dest_addr != dest_addr) {
+		virt_region_free(dest_addr, aligned_dest_addr - dest_addr);
+		dest_addr = aligned_dest_addr;
+	}
 
 	/* Need to make sure this does not step into kernel memory */
 	if (dest_addr < POINTER_TO_UINT(Z_VIRT_REGION_START_ADDR)) {
@@ -492,7 +497,7 @@ void *k_mem_map(size_t size, uint32_t flags)
 	 */
 	total_size = size + CONFIG_MMU_PAGE_SIZE * 2;
 
-	dst = virt_region_alloc(total_size);
+	dst = virt_region_alloc(total_size, CONFIG_MMU_PAGE_SIZE);
 	if (dst == NULL) {
 		/* Address space has no free region */
 		goto out;
@@ -657,9 +662,11 @@ void z_phys_map(uint8_t **virt_ptr, uintptr_t phys, size_t size, uint32_t flags)
 		 "wraparound for physical address 0x%lx (size %zu)",
 		 aligned_phys, aligned_size);
 
+	align_boundary = arch_mmu_virt_align(aligned_phys, aligned_size);
+
 	key = k_spin_lock(&z_mm_lock);
 	/* Obtain an appropriately sized chunk of virtual memory */
-	dest_addr = virt_region_alloc(aligned_size);
+	dest_addr = virt_region_alloc(aligned_size, align_boundary);
 	if (!dest_addr) {
 		goto fail;
 	}
```

And then, on ARM64, `arch_mmu_virt_align()` could look like:

```
size_t arch_virt_align(uintptr_t phys, size_t size)
{
	size_t level_size, mask = phys & size;
	int level;

	for (level = XLAT_LAST_LEVEL; level >= BASE_XLAT_LEVEL; level--) {
		level_size = 1 << LEVEL_TO_VA_SIZE_SHIFT(level);
		if ((mask & (level_size - 1)) == 0) break;
	}

	return level_size;
}
```

On architectures that don't care, a default `arch_mmu_virt_align()` could
simply equivalent to `CONFIG_MMU_PAGE_SIZE`.

*NOTE*: The above was written on a whim and is completely untested.

superna9999 committed:
mmu: get virtual alignment from physical address

On ARM64 platforms, when mapping multiple memory zones with size
not multiple of a L2 block size (2MiB), all the following mappings
will probably use L3 tables.

And a huge mapping will consume all possible L3 tables.

In order to reduce usage of L3 tables, this introduces a new
arch_virt_region_align() optional architecture specific
call to eventually return a more optimal virtual address
alignment than the default MMU_PAGE_SIZE.

This alignment is used in virt_region_alloc() by:
- requesting more pages in virt_region_bitmap to make sure we request
  up to the possible aligned virtual address
- freeing the supplementary pages used for alignment

Suggested-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>

superna9999 committed:
arm64: mmu: implement arch_virt_region_align()

Add the arm64 MMU arch_virt_region_align() implementation used
to return a possible virtual addres alignment in order to
optimize the MMU table layout and possibly avoid using L3 tables
and use some L1 & L3 blocks instead for most of the mapping.

Suggested-by: Nicolas Pitre <npitre@baylibre.com>
Signed-off-by: Neil Armstrong <narmstrong@baylibre.com>
